charms/reactive/endpoints.py

```diff
@@ -139,6 +140,16 @@ def __init__(self, endpoint_name, relation_ids=None):
         self._all_joined_units = None
         self._all_departed_units = None
 
+    def register_triggers(self):
+        """
+        Called once and only once for each named instance of this endpoint,
+        before the endpoint's automatic flags are updated.
+
+        This gives the endpoint implementation a chance to register triggers
+        that will honor changes to the automatically managed flags.
+        """
+        pass
```

Comment:
Yeah, I had realized previously that you couldn't register triggers with endpoints because you couldn't know the endpoint name ahead of time but I didn't think it mattered much.  But then the race condition in the above linked PR / issue came up.  You can't register them in a `@when` type handler because that suffers from the same race condition, and if you register them in `__init__` you might end up with multiple copies (though the instances should be singletons per endpoint name) and you have to deal with passing through the `__init__` params.  This just seemed cleaner.

Non-endpoint code doesn't suffer from this problem because module imports are inherently run before the managed flags and they don't need to know anything ahead of time like the endpoint name.